File: src/Controls/Vns.Erp.Core.Controls/UserControl/Status_LoaiCT.cs
using Microsoft.VisualBasic;
using System;
using System.ComponentModel;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using Vns.Erp.Core.Admin.Domain;
using Vns.Erp.Core.Admin.Service.Interface;
using Vns.Erp.Core;
//using System.ComponentModel;

namespace Vns.Erp.Core.Controls.UserControl
{
    public partial class Status_LoaiCT
    {
        private Vns.Erp.Core.ValueInfo _Value_info = new ValueInfo();

        public Vns.Erp.Core.ValueInfo Value_info
        {
            get { return _Value_info; }
            set { _Value_info = value; }
        }
        //public Vns.Erp.Core.ValueInfo Value_info { get; set; }

        #region "Init Service"
        private void InitService()
        {
            bo_ct_status = (IAppCtStatusService)ObjectFactory.GetObject("AppCtStatusService");
            bo_status_detail = (IKtStatusCtService)ObjectFactory.GetObject("KtStatusCtService");
        }

        #endregion

        #region "Variables"
        IAppCtStatusService bo_ct_status;
        IKtStatusCtService bo_status_detail;
        private KtStatusCt info_detail = new KtStatusCt();
        private Guid DONVI_ID = new Guid();

        private List<Guid> LIST_UG = new List<Guid>();

        private List<KtStatusCt> LISTBIND = new List<KtStatusCt>();

        private Guid _LCT_STATUS = new Guid();
        public Guid LCT_STATUS
        {
            get { return _LCT_STATUS; }
            set { _LCT_STATUS = value; }
        }


        private string _STATUS_UPDATE = "-1";
        public string STATUS_UPDATE
        {
            get { return _STATUS_UPDATE; }
        }

        public void Set_Status(string _ss)
        {
            if (!(HasBindData))
                Loaddata();

            _STATUS_UPDATE = _ss;
            cboStatus.EditValue = _ss;
            if ((_LIST_STATUS_ALL == null) & !(_ss == "-1"))
            {
                _LIST_STATUS_ALL = bo_status_detail.GetAll();
            }
            LayTenTrangThai();
        }


        private bool _CheckStatus = true;
        public bool CheckStatus
        {
            get { return _CheckStatus; }
            set { _CheckStatus = value; }
        }

        /// <summary>
        /// Danh sach cac trang thai trong he thong
        /// </summary>
        /// <remarks></remarks>

        private IList<KtStatusCt> _LIST_STATUS_ALL;
        #endregion

        #region "Common Functions"
        public Status_LoaiCT()
        {
            InitializeComponent();

            if (!this.DesignMode) return;
            InitService();
        }

        private void InitData()
        {
            DONVI_ID = Generals.DON_VI.Id;
            LIST_UG = Vns.Erp.Core.Globals.LIST_USER_GROUP;
            //lblName.Text = ""

        }

        /// <summary>
        /// Lay danh sach tat ca trang thai chung tu
        /// </summary>
        /// <returns></returns>
        /// <remarks></remarks>
        private IList<KtStatusCt> GetListFilter()
        {
            return _LIST_STATUS_ALL;
        }

        /// <summary>
        /// Gan danh sach trang thai duoc phan quyen bind len combobox
        /// </summary>
        /// <remarks></remarks>
        private void GetListStatusFilter()
        {
            try
            {
                IList<KtStatusCt> listFilter = GetListFilter();
                if ((LIST_UG != null))
                {
                    if (LIST_UG.Count > 0)
                    {
                        for (int i = 0; i <= LIST_UG.Count - 1; i++)
                        {
                            Guid group_id = LIST_UG[i];
                            IList<AppCtStatus> dt = bo_ct_status.GetByDVGCT(DONVI_ID, group_id, LCT_STATUS);
                            if (dt.Count > 0)
                            {
                                string ds_status = dt[0].DsStatus;
                                string[] arr = ds_status.Split(';');
                                foreach (string lct_status in arr)
                                {
                                    if ((IsExistedItemInList(LISTBIND, lct_status) == false))
                                    {
                                        KtStatusCt obj = GetDataItem(lct_status);
                                        if ((obj != null))
                                        {
                                            LISTBIND.Add(GetDataItem(lct_status));
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void BindData()
        {
            try
            {
                cboStatus.Properties.DataSource = LISTBIND;
                cboStatus.Properties.ValueMember = "Id";
                cboStatus.Properties.DisplayMember = "Id";
                if (LISTBIND.Count > 0)
                {
                    cboStatus.EditValue = LISTBIND[0].Id;
                }
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void CheckUpdate()
        {
            if (!STATUS_UPDATE.Equals("-1"))
            {
                cboStatus.EditValue = STATUS_UPDATE;
            }
        }

        public bool IsExistedItemInList(List<string> list, string item)
        {
            try
            {
                bool bReturn = false;
                foreach (string info in list)
                {
                    if (info.Equals(item))
                    {
                        bReturn = true;
                    }
                }
                return bReturn;
            }
            catch
            {
                return false;
            }
        }

        public bool IsExistedItemInList(List<KtStatusCt> list, string item)
        {
            try
            {
                bool bReturn = false;
                foreach (KtStatusCt info in list)
                {
                    if (info.Id.Equals(item))
                    {
                        bReturn = true;
                    }
                }
                return bReturn;
            }
            catch
            {
                return false;
            }
        }

        private KtStatusCt GetDataItem(string _Trang_thai)
        {
            foreach (KtStatusCt temp in _LIST_STATUS_ALL)
            {
                if ((temp.Id.ToString() == _Trang_thai))
                {
                    return temp;
                }
            }
            return null;
        }
        #endregion

        #region "Event Handles"
        private void Status_LoaiCT_Load(System.Object sender, System.EventArgs e)
        {
            if ((this.DesignMode))
                return;

            try
            {
                InitService();
                if (!(HasBindData))
                    Loaddata();
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }


        bool HasBindData = false;
        private void Loaddata()
        {
            //Try
            if ((this.DesignMode))
                return;
            if (bo_ct_status == null)
            {
                InitService();
            }
            if ((_LIST_STATUS_ALL == null))
            {
                _LIST_STATUS_ALL = new List<KtStatusCt>(bo_status_detail.GetAll());
                //_LIST_STATUS_ALL ();
            }
            InitData();
            GetListStatusFilter();
            BindData();
            CheckUpdate();
            HasBindData = true;
            //Catch ex As Exception
            //    Message_Error(ex)
            //End Try
        }

        private void cboStatus_EditValueChanged(System.Object sender, System.EventArgs e)
        {
            try
            {
                if (!Generals.Ts_PheDuyetChungTu)
                {
                    _Value_info.ParameterValue = 1;
                    _STATUS_UPDATE = "";
                    return;
                }
                _Value_info.ParameterValue = cboStatus.EditValue;
                _STATUS_UPDATE = _Value_info.ParameterValue.ToString();
                //info_detail = bo_status_detail.GetById(Decimal.Parse(cboStatus.EditValue))
                LayTenTrangThai();
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        #endregion

        //public void selectFirst()
        //{
        //    //cboStatus.Item
        //    //List<KtStatusCt> lst =  cboStatus.Properties.DataSource;
        //    //if (lst.Count != 0)
        //    //{
        //    //    TNCommon.SelectFirst(cboStatus);
        //    //}
        //}

        public bool ChoPhepSuaCT()
        {
            return true;
            //Tam thoi

            bool _bTemp = false;
            foreach (KtStatusCt i in LISTBIND)
            {
                if ((_STATUS_UPDATE == i.Id.ToString()))
                {
                    _bTemp = true;
                    break; // TODO: might not be correct. Was : Exit For
                }
            }
            return (_bTemp & _CheckStatus);
        }

        private string LayTenTrangThai()
        {
            if ((_LIST_STATUS_ALL == null))
            {
                return "";
            }

            foreach (KtStatusCt obj in _LIST_STATUS_ALL)
            {
                if ((STATUS_UPDATE == obj.Id.ToString()))
                {
                    lblName.Text = obj.Ten;
                    return obj.Ten;
                }
            }
            lblName.Text = "";
            return "";
        }
        
    }
}

File: src/Spa/Vns.Erp.Core.Spa.VbSpaGUI/UserControl/Status_LoaiCT.vb
Imports Vns.Erp.Core.Admin.Domain
Imports Vns.Erp.Core.Admin.Service.Interface
Imports Vns.Erp.Core

Public Class Status_LoaiCT
    Public Property Value_info As Vns.Erp.Core.ValueInfo = New Vns.Erp.Core.ValueInfo()

#Region "Init Service"
    Private Sub InitService()
        bo_ct_status = CType(ObjectFactory.GetObject("AppCtStatusService"), IAppCtStatusService)
        bo_status_detail = CType(ObjectFactory.GetObject("KtStatusCtService"), IKtStatusCtService)
    End Sub

#End Region

#Region "Variables"
    Dim bo_ct_status As IAppCtStatusService
    Dim bo_status_detail As IKtStatusCtService
    Private info_detail As New KtStatusCt
    Private DONVI_ID As Guid = New Guid()
    Private LIST_UG As New List(Of Guid)

    Private LISTBIND As New List(Of KtStatusCt)

    Private _LCT_STATUS As Guid = New Guid

    Public Property LCT_STATUS() As Guid
        Get
            Return _LCT_STATUS
        End Get
        Set(ByVal value As Guid)
            _LCT_STATUS = value
        End Set
    End Property

    Private _STATUS_UPDATE As String = "-1"

    Public ReadOnly Property STATUS_UPDATE() As String
        Get
            Return _STATUS_UPDATE
        End Get
    End Property

    Public Sub Set_Status(ByVal _ss As String)
        If Not (HasBindData) Then Loaddata()

        _STATUS_UPDATE = _ss
        cboStatus.EditValue = _ss
        If (_LIST_STATUS_ALL Is Nothing) And Not (_ss = "-1") Then
            _LIST_STATUS_ALL = bo_status_detail.GetAll()
        End If
        LayTenTrangThai()
    End Sub

    Private _CheckStatus As Boolean = True

    Public Property CheckStatus() As Boolean
        Get
            Return _CheckStatus
        End Get
        Set(ByVal value As Boolean)
            _CheckStatus = value
        End Set
    End Property

    ''' <summary>
    ''' Danh sach cac trang thai trong he thong
    ''' </summary>
    ''' <remarks></remarks>
    Private _LIST_STATUS_ALL As List(Of KtStatusCt)

#End Region

#Region "Common Functions"
    Public Sub Status_LoaiCT()
        InitService()
    End Sub

    Private Sub InitData()
        DONVI_ID = Generals.DON_VI.Id
        LIST_UG = Vns.Erp.Core.Globals.LIST_USER_GROUP
        'lblName.Text = ""

    End Sub

    ''' <summary>
    ''' Lay danh sach tat ca trang thai chung tu
    ''' </summary>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Private Function GetListFilter() As List(Of KtStatusCt)
        Return _LIST_STATUS_ALL
    End Function

    ''' <summary>
    ''' Gan danh sach trang thai duoc phan quyen bind len combobox
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub GetListStatusFilter()
        Try
            Dim listFilter As List(Of KtStatusCt) = GetListFilter()
            If Not LIST_UG Is Nothing Then
                If LIST_UG.Count > 0 Then
                    For i As Integer = 0 To LIST_UG.Count - 1
                        Dim group_id As Guid = LIST_UG(i)
                        Dim dt As IList(Of AppCtStatus) = bo_ct_status.GetByDVGCT(DONVI_ID, group_id, LCT_STATUS)
                        If dt.Count > 0 Then
                            Dim ds_status As String = dt(0).DsStatus
                            Dim arr() As String = ds_status.Split(";")
                            For Each lct_status As String In arr
                                If (IsExistedItemInList(LISTBIND, lct_status) = False) Then
                                    Dim obj As KtStatusCt = GetDataItem(lct_status)
                                    If Not (obj Is Nothing) Then
                                        LISTBIND.Add(GetDataItem(lct_status))
                                    End If
                                End If
                            Next
                        End If
                    Next
                End If
            End If
        Catch ex As Exception
            Vns.Erp.Core.Message_Error(ex)
        End Try
    End Sub

    Private Sub BindData()
        Try
            cboStatus.Properties.DataSource = LISTBIND
            cboStatus.Properties.ValueMember = "Id"
            cboStatus.Properties.DisplayMember = "Id"
            If LISTBIND.Count > 0 Then
                cboStatus.EditValue = LISTBIND(0).Id
            End If
        Catch ex As Exception
            Vns.Erp.Core.Message_Error(ex)
        End Try
    End Sub

    Private Sub CheckUpdate()
        If Not STATUS_UPDATE.Equals("-1") Then
            cboStatus.EditValue = STATUS_UPDATE
        End If
    End Sub

    Public Function IsExistedItemInList(ByVal list As List(Of String), ByVal item As String) As Boolean
        Try
            Dim bReturn As Boolean = False
            For Each info As String In list
                If info.Equals(item) Then
                    bReturn = True
                End If
            Next
            Return bReturn
        Catch
            Return False
        End Try
    End Function

    Public Function IsExistedItemInList(ByVal list As List(Of KtStatusCt), ByVal item As String) As Boolean
        Try
            Dim bReturn As Boolean = False
            For Each info As KtStatusCt In list
                If info.Id.Equals(item) Then
                    bReturn = True
                End If
            Next
            Return bReturn
        Catch
            Return False
        End Try
    End Function

    Private Function GetDataItem(ByVal _Trang_thai As String)
        For Each temp As KtStatusCt In _LIST_STATUS_ALL
            If (temp.Id.ToString() = _Trang_thai) Then
                Return temp
            End If
        Next
        Return Nothing
    End Function
#End Region

#Region "Event Handles"
    Private Sub Status_LoaiCT_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        If (Me.DesignMode) Then Return

        Try
            InitService()
            If Not (HasBindData) Then Loaddata()
        Catch ex As Exception
            Vns.Erp.Core.Message_Error(ex)
        End Try
    End Sub

    Dim HasBindData As Boolean = False

    Private Sub Loaddata()
        'Try
        If (Me.DesignMode) Then Return
        If bo_ct_status Is Nothing Then
            InitService()
        End If
        If (_LIST_STATUS_ALL Is Nothing) Then
            _LIST_STATUS_ALL = New List(Of KtStatusCt)
            _LIST_STATUS_ALL.AddRange(bo_status_detail.GetAll())
        End If
        InitData()
        GetListStatusFilter()
        BindData()
        CheckUpdate()
        HasBindData = True
        'Catch ex As Exception
        '    Message_Error(ex)
        'End Try
    End Sub

    Private Sub cboStatus_EditValueChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cboStatus.EditValueChanged
        Try
            If Not Generals.Ts_PheDuyetChungTu Then
                Value_info.ParameterValue = 1
                _STATUS_UPDATE = ""
                Return
            End If
            Value_info.ParameterValue = cboStatus.EditValue
            _STATUS_UPDATE = Value_info.ParameterValue.ToString()
            'info_detail = bo_status_detail.GetById(Decimal.Parse(cboStatus.EditValue))
            LayTenTrangThai()
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

#End Region

    Public Sub selectFirst()
        Dim lst As List(Of KtStatusCt) = cboStatus.Properties.DataSource
        If lst.Count <> 0 Then
            TNCommon.SelectFirst(cboStatus)
        End If
    End Sub

    Public Function ChoPhepSuaCT() As Boolean
        Return True 'Tam thoi

        Dim _bTemp As Boolean = False
        For Each i As KtStatusCt In LISTBIND
            If (_STATUS_UPDATE = i.Id.ToString()) Then
                _bTemp = True
                Exit For
            End If
        Next
        Return (_bTemp And _CheckStatus)
    End Function

    Private Function LayTenTrangThai() As String
        If (_LIST_STATUS_ALL Is Nothing) Then
            Return ""
        End If

        For Each obj As KtStatusCt In _LIST_STATUS_ALL
            If (STATUS_UPDATE = obj.Id.ToString()) Then
                lblName.Text = obj.Ten
                Return obj.Ten
            End If
        Next
        lblName.Text = ""
        Return ""
    End Function
End Class
